Source/WTF/wtf/SequesteredImmortalHeap.h

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (C) 2024-2025 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1.  Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ * 2.  Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/Platform.h>
+
+#if USE(PROTECTED_JIT)
+
+#include <cstddef>
+#include <cstdint>
+#include <mach/mach_vm.h>
+#include <mach/vm_map.h>
+#include <mach/vm_param.h>
+#include <pthread.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+#include <wtf/Assertions.h>
+#include <wtf/Compiler.h>
+#include <wtf/DataLog.h>
+#include <wtf/Lock.h>
+
+#if defined(__has_include)
+#if __has_include(<System/pthread_machdep.h>)
+#include <System/pthread_machdep.h>
+#define HAVE_PTHREAD_MACHDEP_H 1
+#else
+#define HAVE_PTHREAD_MACHDEP_H 0
+#endif
+#else
+#define HAVE_PTHREAD_MACHDEP_H 0
+#endif
+
+namespace WTF {
+
+class SequesteredImmortalHeap {
+    static constexpr bool verbose = true;
+    static constexpr pthread_key_t key = __PTK_FRAMEWORK_JAVASCRIPTCORE_KEY0;
+    static constexpr size_t sequesteredImmortalHeapSlotSize { 16 * 1024 };
+public:
+    static constexpr size_t slotSize { 128 };
+    static constexpr size_t numSlots { 64 };
+
+    enum class AllocationFailureMode {
+        Assert,
+        ReturnNull
+    };
+
+    static SequesteredImmortalHeap& instance()
+    {
+        auto ptr = reinterpret_cast<SequesteredImmortalHeap*>(&s_instance);
+        if (UNLIKELY(!ptr->m_initialized))
+            new (&s_instance) SequesteredImmortalHeap();
```

Comment:
You need to use `std::call_once`
And `m_initialized` initialization order is not well enforced (no store barrier). So let's not do this hand-made singleton and use LazyNeverDestroyed & std::call_once simply.